internal/app/machined/pkg/controllers/etcd/spec.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/talos-systems/talos/pkg/machinery/resources/network"
+)
+
+// SpecController renders manifests based on templates and Spec/secrets.
+type SpecController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *SpecController) Name() string {
+	return "etcd.SpecController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *SpecController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: etcd.NamespaceName,
+			Type:      etcd.ConfigType,
+			ID:        pointer.To(etcd.ConfigID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: network.NamespaceName,
+			Type:      network.HostnameStatusType,
+			ID:        pointer.To(network.HostnameID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: network.NamespaceName,
+			Type:      network.NodeAddressType,
+			ID:        pointer.To(network.FilteredNodeAddressID(network.NodeAddressCurrentID, k8s.NodeAddressFilterNoK8s)),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *SpecController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: etcd.SpecType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *SpecController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		etcdConfig, err := safe.ReaderGet[*etcd.Config](ctx, r, resource.NewMetadata(etcd.NamespaceName, etcd.ConfigType, etcd.ConfigID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting etcd config: %w", err)
+		}
+
+		hostnameStatus, err := safe.ReaderGet[*network.HostnameStatus](ctx, r, resource.NewMetadata(network.NamespaceName, network.HostnameStatusType, network.HostnameID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting hostname status: %w", err)
+		}
+
+		cidrs := make([]string, 0, len(etcdConfig.TypedSpec().ValidSubnets)+len(etcdConfig.TypedSpec().ExcludeSubnets))
+
+		cidrs = append(cidrs, etcdConfig.TypedSpec().ValidSubnets...)
+
+		for _, subnet := range etcdConfig.TypedSpec().ExcludeSubnets {
+			cidrs = append(cidrs, "!"+subnet)
+		}
+
+		// we have trigger on NodeAddresses, but we don't use them directly as they contain
+		// some addresses which are not assigned to the node (like AWS ExternalIP).
+		// we need to find solution for that later, for now just pull addresses directly
+
+		ips, err := net.IPAddrs()
+		if err != nil {
+			return fmt.Errorf("error listing IPs: %w", err)
+		}
+
+		listenAddress := netip.IPv4Unspecified()
+
+		for _, ip := range ips {
+			if ip.To4() == nil {
+				listenAddress = netip.IPv6Unspecified()
```

Comment:
shouldn't we be able to listen on both ipv4 and ipv6?

Reply:
I believe listening on IPv6 implies listening on IPv4.

This issue is that if IPv6 is disabled, we can't listen on IPv6 `::`